armi/runLog.py

```diff
@@ -226,265 +195,346 @@ def setVerbosity(self, levelInput):
+            h = logging.FileHandler(filePath)
+            h.setLevel(logging.WARNING)
+            self.setLevel(logging.WARNING)
 
-    else:
-        raise exceptions.RunLogPromptUnresolvable(
-            "Incorrect CURRENT_MODE for prompting user: {}".format(context.CURRENT_MODE)
-        )
+        form = logging.Formatter(RunLogger.FMT)
+        h.setFormatter(form)
+        self.addHandler(h)
 
+    def _log(self, *args, **kwargs):
+        """wrapper around the standard library Logger._log() method
 
-def warningReport():
-    LOG.warningReport()
+        The primary goal here is to allow us to support the deduplication of warnings.
+        NOTE: All of the *args and **kwargs logic here are mandatory, as the standard library implementation of this
+        method has been changing the number of kwargs between Python v3.4 and v3.9.
+        """
+        # If this is a special "don't duplicate me" string, we need to add that info to the msg temporarily
+        if "single" in kwargs and kwargs["single"]:
+            # if "single" is in the kwards, we need "label", which defaults to None
+            if "label" not in kwargs:
+                kwargs["label"] = label = None
+            else:
+                label = str(kwargs["label"])
+
+            # grab the logging message, default the label to it, if needed
+            msg = str(args[1])
+            label = msg if label is None else label
+
+            # mangle the message, to support our "sometimes we want no duplicates" logging option
+            msg = msg + _SINGLE_STR + label
+
+            # hack the output args
+            largs = list(args)
+            largs[1] = msg
+            args = tuple(largs)
+
+            # the standard library doesn't have single/label, so we must remove them
+            del kwargs["single"]
+            del kwargs["label"]
+        super(RunLogger, self)._log(*args, **kwargs)
+
+    def allowStopDuplicates(self):
+        """helper method to allow us to safely add the deduplication filter at any time"""
+        for f in self.filters:
+            if isinstance(f, DeduplicationFilter):
+                return
+        self.addFilter(DeduplicationFilter())
 
+    def write(self, msg, **kwargs):
+        """the redirect method that allows to do stderr piping"""
+        self.error(msg)
 
-def setVerbosity(level):
-    # convenience function
-    LOG.setVerbosity(level)
+    def flush(self, *args, **kwargs):
+        """stub, purely to allow stderr piping"""
+        pass
 
+    def close(self):
+        """helper method, to shutdown and delete a Logger"""
+        self.handlers.clear()
+        del self
+
+    def getDuplicatesFilter(self):
+        """This object should have a no-duplicates filter. If it exists, find it."""
+        for f in self.filters:
+            if isinstance(f, DeduplicationFilter):
+                return f
+
+        return None
```

Comment:
If we propagate the `single` argument through `extra`, would we still need this?

Reply:
Well, yeah. Because people can add more `Filter` objects to the `Logger`.  That's part of the API, I see no reason to stop people from doing that.